pkg/nmi/server/server.go

```diff
@@ -226,20 +227,53 @@ func (s *Server) msiHandler(logger *log.Entry, w http.ResponseWriter, r *http.Re
 		http.Error(w, msg, http.StatusInternalServerError)
 		return
 	}
-	podns, podname, err := s.KubeClient.GetPodName(podIP)
+	podns, podname, deployment, err := s.KubeClient.GetPodInfo(podIP)
 	if err != nil {
 		logger.Errorf("missing podname for podip:%s, %+v", podIP, err)
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
+
+	// TODO: make it generic for whitelisting of applications
+	micRegEx := regexp.MustCompile(`^mic-*`)
+	micMatch := micRegEx.MatchString(deployment)
+
+	// Request id and request resource extraction are common steps required for
+	// requests from mic as well as other applications.
+	rqClientID, rqResource := parseRequestClientIDAndResource(r)
+
+	// If its mic, then just directly get the token and pass back.
+	if micMatch {
+		var token *adal.Token
+		// UserAssignedIdentity clientID is empty, so we are going to use system assigned MSI
+		if rqClientID == "" {
+			token, err = auth.GetServicePrincipalTokenFromMSI(rqResource)
+		} else { // User assigned identity usage.
+			token, err = auth.GetServicePrincipalTokenFromMSIWithUserAssignedID(rqClientID, rqResource)
+		}
+		if err != nil {
+			logger.Errorf("failed to get service principal token for pod:%s/%s, %+v", podns, podname, err)
+			http.Error(w, err.Error(), http.StatusForbidden)
```

Comment:
Do we want to return the status code returned by metadata ep instead of 403 since we are really just proxying the request here? I think metadata ep will return a `400` or `404`. 403 is not part of retry status codes.

Reply:
Not sure yet, how to extract this info from ada. For time being have put a TODO to handle in a future PR.